### ISSUE TITLE
MTL interop

### ISSUE THREAD

**Issue by arybczak** [tomjaguarpaw/bluefin#5]
Looks like you can't do seamless (i.e. without newtypes) MTL style effects interop, as already evidenced by how [Eff can't have aMonadIO instance](https://github.com/tomjaguarpaw/bluefin/blob/4401ad90ebf5d40261c0013f50cea28c384014f4/bluefin-internal/src/Bluefin/Internal.hs#L43)?

**Comment by tomjaguarpaw** [comment]
That's right, you can't give Bluefin's `Eff` a `MonadWhatever` instance, because the effect is at the value level rather than the constraint level.

**Comment by tomjaguarpaw** [comment]
This approach of passing in multiple effects seems not entirely terrible:

https://github.com/tomjaguarpaw/bluefin/blob/c7b2b1f2680813f402bb4095764beb9daca4fc07/bluefin-internal/src/Bluefin/Internal.hs#L252-L268

**Comment by tomjaguarpaw** [comment]
> Looks like you can't do seamless (i.e. without newtypes) MTL style effects interop

So, for a better answer to your question, you can do MTL style effects interop without (defining your own) newtypes. Whether that's "seamless" or not is up for debate.

**Comment by arybczak** [comment]
Considering that in the "real world" code you will have interleaving of MTL style functions and "native" effects, looks like it's going to be a pain with this approach 🤔

**Comment by tomjaguarpaw** [comment]
> in the "real world" code you will have interleaving of MTL style functions and "native" effects

By '"native" Effects' do you mean `Eff`? If so, that's fine too (see `liftEffStack`):

https://github.com/tomjaguarpaw/bluefin/blob/b9c135448e0820293021b760f5aba5ff58f2560d/bluefin-internal/src/Bluefin/Internal.hs#L260-L272

I'm not going to say this is the most ergonomic thing I've ever seen, but considering it was the first thing I thought of and I worked on it for about an hour, I think it's promising.

**Comment by tomjaguarpaw** [comment]
You could also apply the conversion to individual actions to get the back into `Eff` immediately. I don't know what's going to work out the most ergonomically in practice. In this example everything is done in `Eff` so the `yield` doesn't need to be lifted in any way.

https://github.com/tomjaguarpaw/bluefin/blob/eeb06e37b1ef6a5a5c064a0056b9c835a62f35ad/bluefin-internal/src/Bluefin/Internal.hs#L279-L294

**Comment by arybczak** [comment]
What if you want to call a function `f :: (MonadA m, MonadB m) => m ()`?

**Comment by tomjaguarpaw** [comment]
That one needs the following form (because `mtl` is just `runMTLStyle` fused with `handleMTLWith`):

```.hs
runMTLStyle $ handleMTLWith aHandle $ handleMTLWith bHandle $ f
```

Maybe this form, or some infix operator, would be more ergonomic? It remains to be seen.

```.hs
runMTLStyle (f `handleMTLWith` aHandle `handleMTLWith` bHandle)
```

**Comment by tomjaguarpaw** [comment]
@arybczak, when you first asked this question I assumed that effectful had "seamless (i.e. without newtypes) MTL style effects interop" via instances like:

```.hs
instance Error e :> es => MonadError e (Eff es)
instance State s :> es => MonadState s (Eff es)
instance Reader r :> es => MonadReader r (Eff es)
```

but now that I've got round to looking, a year later, it seems that this is not actually the case. What is the story for "seamless" MTL interop in effectful? For example, suppose I want to call a function

```.hs
f :: (MonadState s m, MonadError e m) => m ()
```

in effectful.
